client/state/data-layer/wpcom/login-2fa/index.js

```diff
@@ -20,77 +20,83 @@ import {
-			auth_type: authType,
-			two_step_nonce: getTwoFactorAuthNonce( store.getState(), authType ),
-			remember_me: getRememberMe( store.getState() ),
-			two_step_push_token: getTwoFactorPushToken( store.getState() ),
-			client_id: config( 'wpcom_signup_id' ),
-			client_secret: config( 'wpcom_signup_key' ),
-		} ).then( () => {
-			store.dispatch( { type: TWO_FACTOR_AUTHENTICATION_PUSH_POLL_COMPLETED } );
-		} ).catch( error => {
-			store.dispatch( {
-				type: TWO_FACTOR_AUTHENTICATION_UPDATE_NONCE,
-				nonceType: 'push',
-				twoStepNonce: error.response.body.data.two_step_nonce
-			} );
-			return Promise.reject( error );
-		} );
+	store.dispatch(
+		http( {
+			url: 'https://wordpress.com/wp-login.php?action=two-step-authentication-endpoint',
+			method: 'POST',
+			headers: [
+				[ 'Content-Type', 'application/x-www-form-urlencoded' ],
+			],
+			withCredentials: true,
+			body: {
+				user_id: getTwoFactorUserId( store.getState() ),
+				auth_type: authType,
+				two_step_nonce: getTwoFactorAuthNonce( store.getState(), authType ),
+				remember_me: getRememberMe( store.getState() ),
+				two_step_push_token: getTwoFactorPushToken( store.getState() ),
+				client_id: config( 'wpcom_signup_id' ),
+				client_secret: config( 'wpcom_signup_key' ),
+			},
+		},
+		action
+	) );
 };
 
+const receivedTwoFactorPushNotificationApproved = ( { dispatch } ) =>
+	dispatch( { type: TWO_FACTOR_AUTHENTICATION_PUSH_POLL_COMPLETED } );
+
 /***
- * Polling the login API for the status of the push notification.
- * The polling will stop on success or when store's polling progress
- * state changes to `false`
+ * Receive error from the two factor push notification status http request
  *
- * @param {Object}   store  Global redux store
+ * @param {Object}	store  Global redux store
+ * @param {Object}	action dispathced action
+ * @param {Object}	error the error object
  */
-const doAppPushPolling = store => {
-	let retryCount = 0;
-	const retry = () => {
-		// if polling was stopped or not in progress - stop
-		if ( ! getTwoFactorPushPollInProgress( store.getState() ) ) {
-			return;
-		}
+const receivedTwoFactorPushNotificationError = ( store, action, next, error ) => {
+	const twoStepNonce = get( error, 'response.body.data.two_step_nonce' );
 
-		retryCount++;
-		setTimeout(
-			() => doAppPushRequest( store ).catch( retry ),
-			( POLL_APP_PUSH_INTERVAL_SECONDS + Math.floor( retryCount / 10 ) ) * 1000 // backoff lineary
-		);
-	};
+	if ( ! twoStepNonce ) {
+		store.dispatch( { type: TWO_FACTOR_AUTHENTICATION_PUSH_POLL_STOP } ); //
+		throw new Error( "Two step nonce wasn't present on the response from polling endpoint, unable to continue" );
+	}
+
+	store.dispatch( {
+		type: TWO_FACTOR_AUTHENTICATION_UPDATE_NONCE,
+		nonceType: 'push',
+		twoStepNonce,
+	} );
 
 	if ( getTwoFactorPushPollInProgress( store.getState() ) ) {
-		doAppPushRequest( store ).catch( retry );
+		setTimeout(
+			// eslint-disable-next-line no-use-before-define
```

Comment:
The current requirement is to have polling active until the user navigates to another screen, then the polling stops.

Currently it's done with that [component](https://github.com/Automattic/wp-calypso/blob/master/client/blocks/login/two-factor-authentication/push-notification-approval-poller.jsx):
```js
class PushNotificationApprovalPoller extends Component {
        ...

	componentDidMount() {
		this.props.startPollAppPushAuth();
	}

	componentWillUnmount() {
		this.props.stopPollAppPushAuth();
	}

	componentWillReceiveProps( nextProps ) {
		if ( ! this.props.pushSuccess && nextProps.pushSuccess ) {
			this.props.onSuccess();
		}
	}

        ...
}

export default connect(
	( state ) => ( {
		pushSuccess: getTwoFactorPushPollSuccess( state ),
	} ),
	{
		startPollAppPushAuth,
		stopPollAppPushAuth,
	}
)( PushNotificationApprovalPoller );
``` 

So we need at least 2 actions, one for the polling to stop, and another one for it to stop.
so we can have something like:
`TWO_FACTOR_AUTHENTICATION_PUSH_POLL` - will start the polling, and will get the successful response to the store when it'll propagate after the data layer
`TWO_FACTOR_AUTHENTICATION_PUSH_POLL_CANCEL` - will cancel polling

So something like:
```js
...
export default {
	[ TWO_FACTOR_AUTHENTICATION_PUSH_POLL ]: [ ( store, action ) => {
		// if it's a "fresh" action, reset cancel flag
		if ( ! get( action, 'meta.retryCount', false ) && cancelPolling ) {
				cancelPolling = false;
		}

		return dispatchRequest(
			requestTwoFactorPushNotificationStatus,
			receivedTwoFactorPushNotificationApproved,
			receivedTwoFactorPushNotificationError,
		);
	} ],
	[ TWO_FACTOR_AUTHENTICATION_PUSH_CANCEL ]: [ () => {
		cancelPolling = true;
	} ],
};
...
```
more complete diff: https://gist.github.com/yurynix/c24fa7c7e9ade4c167e4f026ab6a8e67

But then the problem is that the original `TWO_FACTOR_AUTHENTICATION_PUSH_POLL` is going through the store and updating the reducer that the operation is complete, so 
i'll need another action that indicates the completion of the operation, so I'm back to where I started - 3 actions.

I feel that this approach doesn't brings huge benefits ATM, so I prefer to merge as is, unless you strongly feel about it.